dashboard/db/migrate/20210625170655_add_script_id_and_deleted_at_to_channel_tokens.rb

```diff
@@ -0,0 +1,8 @@
+class AddScriptIdAndDeletedAtToChannelTokens < ActiveRecord::Migration[5.2]
+  def change
+    add_column :channel_tokens, :script_id, :integer
+    add_column :channel_tokens, :deleted_at, :datetime
+    add_index :channel_tokens, [:storage_id, :level_id, :script_id, :deleted_at], unique: true, name: 'index_channel_tokens_unique' # the generated name was too long
```

Comment:
Not sure if it matters for our use case, but we might want `script_id` ahead of `level_id` in the index?  (Maybe @sureshc has some advice here.)

Reply:
adding the script_id at the end seems safest. if you put it at the beginning, these queries would become unindexed:
* https://github.com/code-dot-org/code-dot-org/blob/997a27899ed58ab1ad7853af0f2f6bf9ca29143c/dashboard/app/models/channel_token.rb#L40
* https://github.com/code-dot-org/code-dot-org/blob/997a27899ed58ab1ad7853af0f2f6bf9ca29143c/dashboard/app/models/channel_token.rb#L49-L50

For background, see https://www.mysqltutorial.org/mysql-index/mysql-composite-index:

> Notice that if you have a composite index on (c1,c2,c3), you will have indexed search capabilities on one the following column combinations:
>
> (c1)
> (c1,c2)
> (c1,c2,c3)

because the table is so big, I would assume that we can't afford to have unindexed queries for any period of time, even if you plan to add script_id to those queries the same day (or even during the same deploy).

Reply:
Oh yikes, good catch.  Should we add a new index and then remove the old one after the above references have been fixed then?  (I'm assuming that everywhere we query by level id, we want to add script id.)

Reply:
not that I'm aware of -- the problems I described would only occur if you were to move script_id to the beginning of the list of indexed fields. the PR as it is currently written looks good to me.

if you did need to put script_id first, then that approach you're describing would make sense.

Reply:
in case it helps... 

>because the table is so big, I would assume that we can't afford to have unindexed queries for any period of time, even if you plan to add script_id to those queries the same day (or even during the same deploy). 

i was trying to say that if you have a backward incompatible migration, it is NOT sufficient to update the code in the same PR to start using the new index. the reason is that we would have database problems between the time that the db migration took place and the time that the frontend servers were updated to use the new code.

Reply:
Logically, I think it makes more sense for script_id to be ahead of level_id, but perhaps it doesn't make a difference for our current usage patterns (because e.g., we don't query for channel_tokens by script_id).  Not sure if it's worth making this more future-proof now.

Btw, should we add an index just on `deleted_at` so we can quickly find all deleted entries?  (Not sure how our cleanup logic works / will work.)

Reply:
I would suggest optimizing for the specific set of queries that you'll need in order to complete the current round of "reset progress" work. perhaps identifying the specific queries you plan to use would be a good next step.

if you want to rearrange the order of the indexed columns and you do not foresee a downside to doing so (based on specific query needs), that's also fine with me. In that case, I think you have the right idea of how to accomplish that.

for the deleted_at column in particular, I did notice that this column is sometimes indexed separately and sometimes indexed jointly with other columns: 
![Screen Shot 2021-06-28 at 2 22 41 PM](https://user-images.githubusercontent.com/8001765/123706543-4187c480-d81d-11eb-983f-19663c2ca89e.png) I would imagine that having a separate deleted_at column would solve the problem for you, since I would expect the database engine to be able to use multiple indexes during a single query. unfortunately, I don't know what the exact tradeoffs are between using a single compound index (when possible) vs using two separate indexes.

Whatever you decide, I think @sureshc will need to chime in on how to ship these migrations safely, due to the size of the table.